flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/CompoundReducingUpsertWriterStateSerializer.java

```diff
@@ -0,0 +1,123 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.table;
+
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.table.data.RowData;
+
+import javax.annotation.Nullable;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+class CompoundReducingUpsertWriterStateSerializer<WriterState>
+        implements SimpleVersionedSerializer<CompoundReducingUpsertWriterState<WriterState>> {
+
+    private final TypeSerializer<RowData> rowDataTypeSerializer;
+    @Nullable private final SimpleVersionedSerializer<WriterState> wrappedStateSerializer;
+
+    CompoundReducingUpsertWriterStateSerializer(
+            TypeSerializer<RowData> rowDataTypeSerializer,
+            @Nullable SimpleVersionedSerializer<WriterState> wrappedStateSerializer) {
+        this.wrappedStateSerializer = wrappedStateSerializer;
+        this.rowDataTypeSerializer = checkNotNull(rowDataTypeSerializer);
+    }
+
+    CompoundReducingUpsertWriterStateSerializer(TypeSerializer<RowData> rowDataTypeSerializer) {
+        this(rowDataTypeSerializer, null);
+    }
+
+    @Override
+    public int getVersion() {
+        return 1;
+    }
+
+    @Override
+    public byte[] serialize(CompoundReducingUpsertWriterState<WriterState> state)
+            throws IOException {
+        try (final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                final DataOutputViewStreamWrapper out = new DataOutputViewStreamWrapper(baos)) {
+            final List<WriterState> wrappedStates = state.getWrappedStates();
+            if (wrappedStateSerializer != null) {
+                out.writeInt(wrappedStateSerializer.getVersion());
+                out.writeInt(wrappedStates.size());
+                for (final WriterState wrappedState : wrappedStates) {
+                    final byte[] serializedWrappedState =
+                            wrappedStateSerializer.serialize(wrappedState);
+                    out.writeInt(serializedWrappedState.length);
+                    out.write(wrappedStateSerializer.serialize(wrappedState));
+                }
```

Comment:
Extract to `SimpleVersionedSerialization`?

Reply:
Thanks for the hint, I did not know that such a class exists but I think does not help much here. The `SimpleVersionedSerialization` always writes the version in front of the serialized which in this cases leads to a lot of duplicated version writes.
In general, the `SimpleVersionedSerialization` seems not to be great for writing `List<>` data.

Reply:
That's why I was hoping if we could add it to `SimpleVersionedSerialization`.

Reply:
Now I understand your point I'll have a look.